app/master/cluster_master.py

```diff
@@ -179,7 +179,7 @@ def _disconnect_slave(self, slave):
         """
         # Mark slave dead. We do not remove it from the list of all slaves. We also do not remove it from idle_slaves;
         # that will happen during slave allocation.
-        slave.is_alive = False
+        slave.set_is_alive(False)
```

Comment:
@wjdhollow Will, I see your point about potentially creating a third in-between state, but I feel like it might be better to make the slaves fail fast and just shut down completely when something goes wrong instead of going into zombie limbo. Or did you mean the zombie state would just be a temporary/transitional state?

@tjlee0909 I agree with your last sentence, TJ. I think the behavior we should strive to make the slave conform to is that the slave (from the master's perspective) is either: (1) responsive / alive / completely functional, or (2) unresponsive / dead / not running. 

Right now though, when the slave is shutting down, it will disconnect from the master before its web server goes down, so there **would** be a window where the slave would still have a responsive API but would not expect to receive requests from the master. (The slave is doing build teardown commands during that time, so that window could last an indefinite duration.)

That window is not a risk right now. Specifically, on the master side, your change only actively pings the slave in the slave allocation loop, and the loop only acts on slaves in the `_idle_slaves` queue. The only way a slave gets put back into the idle queue is if it sends a notification to the master that it is idle (via `ClusterMaster.handle_slave_state_update`). Essentially, a slave has to specifically tell the master that it wants to be put back into the idle queue, and that won't happen if the slave has disconnected.

Obvs, that doesn't mean the scenario you mentioned couldn't come up as an issue in the future though. One solution would be to have a specific API endpoint on the slave by which the slave could report its own dead/alive status. (And ideally the slave would always respond with "alive" except during shutdown.) That would decouple the responsiveness of the web server from the internal state of the slave. 

Ugh, sorry long response. :sleeping: